### PR TITLE
feat: add support for custom offline maps

### LINKED ISSUE #243: Use map serving functionality of mapeo core
## Description

As part of the work for https://github.com/digidem/mapeo-core-next/issues/437, we made online and offline (both custom + fallback) map serving the responsibility of the http server that's used in tandem with mapeo core. This feature has not yet been integrated into the app yet, but doing so would allow the frontend to avoid any logic around map style url resolving based on network states.

## Tasks

- backend
  - [x] set up fastify plugins for serving maps
  - [x] add default offline map (from https://github.com/digidem/mapeo-offline-map?) as an asset that's included in the app

- frontend
  - [x] update the style url used any MapView instances to use the url returned by `MapManager.getMapStyleJsonUrl()`

### PR DESCRIPTION
**Pull request by achou11**
Closes #243 

Notes:

- decision tree for which map's style.json to serve (EDIT: when map initially loads/renders):
  
    image

Remaining items:

- [X] on my Pixel 2 Android 11, zooming out eventually causes the app to crash. have only tried this with one example map so not sure if it's specific to this map or a general issue. i get the following some cryptic logs from adb but not sure how to debug:

    ```sh
    NODEJS-MOBILE: exiting due to SIG_DFL handler for signal 11
    ```
    
    ```sh
    backtrace:
      #00 pc 000000000004e0f4  /apex/com.android.runtime/lib64/bionic/libc.so (abort+180) (BuildId:   
      #01 pc 00000000000050b8  /system/bin/app_process64 (art::SignalChain::Handler(int, siginfo*, void*)+1120) (BuildId:   
      #02 pc 00000000000008b0  [vdso] (__kernel_rt_sigreturn)
      #03 pc 0000000000118c3c  [anon:scudo:primary]
    ```
    
    One thing to try next: upload same map to Mapeo Mobile and see if same crash happens.
    
    EDIT: I'm no longer getting the crash for mysterious reasons...

- [ ] ~attempting to test this on my Pixel 6 Android 14 is very difficult due to storage permission changes introduced in Android 13+14. Basically almost impossible without using adb and even then, it requires some complicated workarounds that I have yet to figure out (think other members of our team have done it before).~ EDIT: will not test on pixel 6 for now

### CODE REVIEW

**Comment by gmaclennan** [comment]
Some initial ideas for debugging (you may have already tried): turn on logging for fastify so you can see what requests are being handled, so you can trace back what is happening up to the crash.

**Comment by achou11** [comment]
For reasons unknown, I'm no longer getting the crash. Didn't do anything special...

**Review by ErikSin — CHANGES_REQUESTED** [review]
The use of useMapStyle() feels unnecessary with the tanstack query.

I think it would be better to use a suspense query (The map is the main feature, so i think it makes sense to block the entire app while it is loading)

And then we could create a custom hook for invalidating the query when the user goes offline/online

```ts
// custom hook
function useInvalidateMapUrlQuery(){
     const {isInternetReachable} = useNetInfo();
     const queryClient = useQueryClient()
     useEffect(()=>{
          queryClient.invalidateQuery({queryKey:[MAP_STYLE_URL_KEY]})
     },[queryClient, isInternetReachable])
}

// inside map screen:
const MapScreen = () => {
     const {data:mapStyleUrl} = useMapStyleUrl() //this is a suspense query, so no loading needs to be taken into account
     useInvalidateMapUrlQuery()

     // rest of component
}

```

The code feels a little cleaner if we are just utilizing tanstack query.

The other thing that we should maybe do is only change to the offline map if there isnt an online map already loaded. Aka, if the user is online and has a map loaded and then goes offline, don't reload the map with offline one, just keep the online one (even though we wont have any zoom) The online map at the farthest zoom feels better than using an offline map.

Basically the only time we should serve an offline map is when they open the app initally without any access to the internet.

**Comment by achou11** [comment]
> The other thing that we should maybe do is only change to the offline map if there isnt an online map already loaded. 

not entirely sure which map you're referring to when saying "offline", but for custom background maps, the legacy behavior is that if there's a custom background map loaded (e.g. a map that exists in `files/styles/default` folder in android app storage), then that takes first priority, even if you're online. 

Regarding the proposed approach for adjusting the setup for the query, your suggestion as is will not work because the URL returned by the api is stable during the lifetime of the backend server. Basically it will always point to `http:localhost:<some_port>/maps/style.json`. Only invalidating the query will return the same URL, but what we need is a way for the style.json resource to be re-requested by the map library, which is what the current solution of using a search param does in this case.

Agree that the query can probably be switched to a suspense query, but the `useMapStyle()` hook as it's implemented otherwise still feels appropriate

File: src/backend/src/app.js [hunk 1]
```diff
@@ -47,27 +57,46 @@ process.on('exit', (code) => {
...
+  fastify.register(MapeoOfflineFallbackMapFastifyPlugin, {
+    prefix: 'fallback',
+    styleJsonPath: join(fallbackMapPath, 'style.json'),
+    sourcesDir: join(fallbackMapPath, 'dist'),
+  })
+  fastify.register(MapeoMapsFastifyPlugin, {
+    prefix: 'maps',
+    defaultOnlineStyleUrl: DEFAULT_ONLINE_MAP_STYLE_URL,
+  })
+
```
[the remark this answers is not in the document]
Comment (achou11) [reply src/backend/src/app.js:99]: it's to add maps functionality to the fastify instance that we use for serving things over http. the functionality is separate into 3 plugins:

- static: provides ability to statically serve custom offline maps
- fallback: provides ability to statically serve a directory that acts as the fallback map (which are set up differently that custom offline maps). this is mostly to accommodate for the fallback map structure that's used by Mapeo Mobile.
- maps: currently acts as the main resolver for the user-facing API e.g. resolving the appropriate style.json based on existence of static/fallback maps. in the future, this will probably hold other functionality related to features that are more similar to the Map Manager in Mapeo Mobile.

**Comment by achou11** [comment]
> The other thing that we should maybe do is only change to the offline map if there isnt an online map already loaded. Aka, if the user is online and has a map loaded and then goes offline, don't reload the map with offline one, just keep the online one (even though we wont have any zoom) The online map at the farthest zoom feels better than using an offline map.

Agree that there's probably improvements to the UX around this stuff. but this PR is mostly focused on porting over the existing behavior from Mapeo Mobile, which is described by the decision tree above

**Comment by achou11** [comment]
I could couple the net info check into the query hook and that would probably work, but not sure if there's any notable advantage to doing that in this case e.g.

```ts
export function useMapStyleUrl() {
  const {isInternetReachable} = useNetInfo();
  const api = useApi();

  return useSuspenseQuery({
    queryKey: [MAP_STYLE_URL_KEY],
    queryFn: async () => {
      const url = await api.getMapStyleJsonUrl();

      return isInternetReachable ? url : url + '?offline=true';
    },
  });
}
```

**Comment by achou11** [comment]
upon reading the docs for query invalidation, i think what you proposed should work. i was worried that react query wouldn't actually refetch even on invalidation, but seems like it will by default, which suffices for our needs at the moment. There may be a time where having a search parameter like `offline=true` could be helpful, but as I mentioned in an earlier comment, the backend doesn't do anything with that for now.

**Comment by achou11** [comment]
hm switching to use a suspense query doesn't seem to work for me 🤔 end up with just a blank white space where the map's supposed to be. maybe due to the mapping library not playing well with suspense, or where our suspense boundary is?

image

**Comment by achou11** [comment]
> > The other thing that we should maybe do is only change to the offline map if there isnt an online map already loaded.
> 
> not entirely sure which map you're referring to when saying "offline", but for custom background maps, the legacy behavior is that if there's a custom background map loaded (e.g. a map that exists in `files/styles/default` folder in android app storage), then that takes first priority, even if you're online. if you're talking about switching from online tiles to the offline fallback map, that would require fixes on the backend side and probably better to do as a follow-up if needed.
> 
> Regarding the proposed approach for adjusting the setup for the query, your suggestion as is will not work because the URL returned by the api is stable during the lifetime of the backend server and will not change based on internet connectivity. Basically it will always point to `http://localhost:<some_port>/maps/style.json`. Only invalidating the query will return the same URL, but what we need is a way for the _**style.json resource to be re-requested by the map library**_, which is what the current solution of using a search param does in this case. the search param is ignored by the server on the backend - it's basically used to tell the mapping library to re-request the style.json resource.
> 
> Agree that the query can probably be switched to a suspense query, but the `useMapStyle()` hook as it's implemented otherwise still feels appropriate

I misspoke on the first part here. Without any custom basemap, I thought the behavior when going from online to offline was that the fallback offline map would immediately be displayed in the app. Based on testing Mapeo Mobile, this is incorrect, and what you suggested is what actually happens in this case (again, wasn't sure what you mean by "offline"). Seems like the fallback offline map is never _dynamically_ displayed i.e. when network state changes while app is active. It seems that it's only ever displayed when the map initially renders.

This is an easy fix in this PR, as it basically means I can get rid of the network-related parts of the changes, which will simplify the PR as a side effect!

**Review by ErikSin — APPROVED** [review]
Great, everything is working 

This is out of scope of this PR, but we should check when the user goes back online and serve that online map without forcing them to close the app.

**Comment by ErikSin** [comment]
For the posterity I didn't actually test this with a custom map...For the MVP release that doesn't feel like a priority so I will make note it for the QA testers

**Comment by achou11** [comment]
> For the posterity I didn't actually test this with a custom map...For the MVP release that doesn't feel like a priority so I will make note it for the QA testers

also for posterity, i tested this on my Pixel 2 Android 11 using some maps from the Mapeo testing kit (which we have in Google Drive)
